modules/study/src/main/Socket.scala

```diff
@@ -41,12 +42,35 @@ private final class Socket(
     lilaBus.unsubscribe(self)
   }
 
+  def sendStudyJoin(userId: User.ID) {
+    lightStudyCache.get(studyId.value) foreach { studyOption =>
+      studyOption foreach { study =>
+        val contributor = study.contributors.contains(userId)
+        context.system.lilaBus.publish(lila.hub.actorApi.study.StudyJoin(userId, studyId.value, contributor, study.isPublic), 'study)
+      }
+    }
+  }
+
+  def sendStudyQuit(userId: User.ID) {
+    lightStudyCache.get(studyId.value) foreach { studyOption =>
+      studyOption foreach { study =>
+        val contributor = study.contributors.contains(userId)
+        context.system.lilaBus.publish(lila.hub.actorApi.study.StudyQuit(userId, studyId.value, contributor, study.isPublic), 'study)
+      }
+    }
+  }
+
   def receiveSpecific = ({
 
-    case SetPath(pos, uid) => notifyVersion("path", Json.obj(
-      "p" -> pos,
-      "w" -> who(uid).map(whoWriter.writes)
-    ), noMessadata)
+    case SetPath(pos, uid) =>
+      uidToUserId(uid) match {
```

Comment:
also `SetPath` can happen hundreds of time per minute, in a single study. Make sure it cannot trigger something costly (like `fetchFollowers`)